include/hybridmap/AodScheduler.hpp

```diff
@@ -0,0 +1,302 @@
+//
+// This file is part of the MQT QMAP library released under the MIT license.
+// See README.md or go to https://github.com/cda-tum/qmap for more information.
+//
+
+#pragma once
+
+#include "Definitions.hpp"
+#include "QuantumComputation.hpp"
+#include "cstdint"
+#include "hybridmap/NeutralAtomArchitecture.hpp"
+#include "hybridmap/NeutralAtomDefinitions.hpp"
+#include "hybridmap/NeutralAtomUtils.hpp"
+#include "operations/AodOperation.hpp"
+#include "operations/OpType.hpp"
+#include "utility"
+#include "vector"
+
+namespace qc {
+// Possible types two Move combination can be combined to
+enum class ActivationMergeType : uint8_t { Impossible, Trivial, Merge, Append };
+
+/**
+ * @brief Class to schedule AOD movements on a neutral atom architecture
+ * @details The scheduler takes a quantum circuit containing abstract move
+ * operations and tries to merge them into parallel AO movements. It also
+ * manages the small offset movements required while loading or unloading of
+ * AODs.
+ */
+class AodScheduler {
+protected:
+  /**
+   * @brief Struct to store information about specific AOD activations.
+   * @details It contains:
+   * - the offset moves in x and y direction
+   * - the actual moves
+   */
+  struct AodActivationHelper {
+    /**
+     * @brief Describes a single AOD movement in either x or y direction
+     * @details It contains:
+     * - the initial position of the AOD
+     * - the delta of the AOD movement
+     * - the size of the offset of the AOD movement
+     */
+    struct AodMove {
+      // start of the move
+      uint32_t init;
+      // need offset move to avoid crossing
+      int32_t offset;
+      // delta of the actual move
+      fp delta;
+
+      AodMove(uint32_t init, fp delta, int32_t offset)
+          : init(init), offset(offset), delta(delta) {}
+    };
+    /**
+     * @brief Manages the activation of an atom using an AOD.
+     * @details The same struct is used also to deactivate the AOD, just
+     * reversed.
+     */
+    struct AodActivation {
+      // first: x, second: delta x, third: offset x
+      std::vector<AodMove*> activateXs;
+      std::vector<AodMove*> activateYs;
+      std::vector<AtomMove> moves;
+
+      AodActivation(const AodMove& activateXs, const AodMove& activateYs,
+                    const AtomMove& move)
+          : activateXs({new AodMove(activateXs)}),
+            activateYs({new AodMove(activateYs)}), moves({move}) {}
+      AodActivation(const Dimension dim, const AodMove& activate,
+                    const AtomMove& move)
+          : moves({move}) {
+        if (dim == Dimension::X) {
+          activateXs = {new AodMove(activate)};
+        } else {
+          activateYs = {new AodMove(activate)};
+        }
+      }
+
+      [[nodiscard]] std::vector<AodMove*> getActivates(Dimension dim) const {
+        if (dim == Dimension::X) {
+          return activateXs;
+        }
+        return activateYs;
+      }
+    };
```

Comment:
The use of `new` here without corresponding `delete` calls screams memory leaks. Consider using smart pointers (probably `std::shared_ptr`) instead or to replace the pointers by pure values in general (assuming the pointer semantics are not needed).

Reply:
Tried to not use pointers at all. Failed. Tried again. Failed again. Now I switched to shared_ptr

Reply:
I can also give this a try tomorrow and see whether I can get it to work without pointers 😌